### PR TITLE
Normal Model

### PR DESCRIPTION
**Pull request by ilan-gold**
Opening as a draft because I have a few questions:

1.  What should be done about the unnecessary parameters in the `Estimator` class?  We don't need to initialize anything because the model is fit in closed form
2. We have one scale parameter for all observations collectively for a given model.  This has some weird side-effects
 a. What do we use in the final likelihood? Since we could recover a scale estimate for each observation (i.e `self.scale`) I'm not sure what to do.  Of course, we have the closed-form estimation of the variance from the residuals so that is what I used here but it does seem a bit strange to not use the fitted values in the final likelihood.  But then the final likelihood would be of a different form from the likelihood that was used as the basis for estimation (i.e many scale values vs. one)
 b. Is what I'm doing even correct?  Judging by the manuscript, I think the answer is "yes" but I'm not sure.
3. Any good ways to test this for correctness?  Our accuracy tests don't work as of now.  Looking at the errors for normality? Relatedly, and in good news, is there is no OOM error (tested on 20000 features)
4. Is there any desire to add tests for checking the shape of outputs if not their correctness? I checked by hand but this seems testable.
5. In `closedform_glm_scale` and `closedform_glm_mean` we do this grouping step before computing group-wise means/moments.  This seems like something we don't want to do here, but I'm not sure since the functions were already written so presumably they are meant to be used?  But maybe they are just an artifact/only necessary for IWLS initialization.  It seems like for actual training (as opposed to initialization) we would want to leave this grouping choice up to the consumer a la pseudo-bulk vs. single-cell in DE testing.

Thanks

### CODE REVIEW

**Comment by davidsebfischer** [comment]
1. ignore or raise flags if not meaningful?
3. accuracy tests > maybe tune the hyperparameters of simulation and test for each distributition for now? this depends on a stoachsticity in simulation and ability of estimator to recover MLE of simulation
4. shape tests > yes

File: batchglm/models/glm_norm/utils.py [hunk 1]
```diff
@@ -71,3 +42,74 @@ def compute_scales_fun(variance, mean):
...
+                (
+                    "There is no need for closed form location model initialization"
+                    "because it is already closed form - falling back to zeros"
+                )
+            )
+        init_theta_location = np.zeros([model.num_loc_params, model.num_features])
+    elif init_location_str == "standard":
+        overall_means = np.mean(model.x, axis=0)  # directly calculate the mean
+        init_theta_location = np.zeros([model.num_loc_params, model.num_features])
+        init_theta_location[0, :] = np.log(overall_means)
```
Comment (picciama) [inline batchglm/models/glm_norm/utils.py:83]: why is this log? Isn't the linker function the identity here? Because we are assume a normal distribution?

File: batchglm/models/glm_norm/utils.py [hunk 2]
```diff
@@ -1,47 +1,18 @@
 import logging
-from typing import Union
+from typing import Union, Tuple
 
 import numpy as np
 import scipy.sparse
+import dask
 
 from .external import closedform_glm_mean, closedform_glm_scale
```
Comment (picciama) [inline batchglm/models/glm_norm/utils.py:8]: closedform_glm_mean isn't used.

File: batchglm/train/numpy/glm_norm/external.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import batchglm.utils.data as data_utils
+from batchglm import pkg_constants
+from batchglm.models.base_glm.utils import closedform_glm_mean, closedform_glm_scale
```
Comment (picciama) [inline batchglm/train/numpy/glm_norm/external.py:3]: `closedform_glm_mean` isn't used.

File: batchglm/models/glm_norm/utils.py [hunk 4]
```diff
@@ -71,3 +42,74 @@ def compute_scales_fun(variance, mean):
...
+            constraints=model.constraints_scale,
+            size_factors=model.size_factors,
+            groupwise_means=groupwise_means
+        )
+    elif init_scale_str == "all_zero":
+        init_theta_scale = np.zeros([model.num_scale_params, model.x.shape[1]])
+    else:
+        raise ValueError("init_scale string %s not recognized" % init_scale_str)
+
+    return init_theta_location, init_theta_scale, True, True
```
Comment (picciama) [inline batchglm/models/glm_norm/utils.py:115]: why would the last two return values, i.e. `train_loc` and `train_scale` be true? The norm model is computed in closed form so shouldn't that be `False`? If the initialization isn't in closed form but e.g. `all_zeros`, then only do we need to set these two values to `True` I guess?

File: batchglm/train/numpy/glm_norm/estimator.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import logging
+
+import numpy as np
+
+from .external import EstimatorGlm, Model
+from .model_container import ModelContainer
+from ....models.glm_norm.utils import init_par
```
Comment (picciama) [inline batchglm/train/numpy/glm_norm/estimator.py:7]: add the `init_par` import to `external.py`

File: batchglm/train/numpy/glm_norm/estimator.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            - str:
+                * "auto": automatically choose best initialization
+                * "random": initialize with random values
+                * "standard": initialize with zeros
+                * "closed_form": try to initialize with closed form
+            - np.ndarray: direct initialization of 'b'
+        :param quick_scale: bool
+            Whether `scale` will be fitted faster and maybe less accurate.
+            Useful in scenarios where fitting the exact `scale` is not absolutely necessary.
+        :param dtype: Numerical precision.
```
Comment (picciama) [inline batchglm/train/numpy/glm_norm/estimator.py:47]: please check the documented values for `init_location` and `init_scale`. It seems not all of them are implemented in `init_par`.

File: batchglm/train/numpy/glm_norm/estimator.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        self,
+        **kwargs,
+    ):
+        model = self._model_container.model
+        if self._train_loc:
+            theta_location, _, _, _ = np.linalg.lstsq(model.design_loc, model.x)
+            self._model_container.theta_location = theta_location
+        self._train_loc = False
+        super().train(**kwargs)
+        self._train_loc = True
```
Comment (picciama) [inline batchglm/train/numpy/glm_norm/estimator.py:77]: why is `train_loc` and `train_scale` handled here and the parent training intercepted? And why is the training specific `theta_location` in `model_container` overwritten? Can you add a docstring here if this is absolutely needed? Also, why would you call the parent training after `train_loc` is set to `False`, then afterwards set it to `True`? Shouldn't it be the other way round?

File: batchglm/train/numpy/glm_norm/model_container.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        return ll
+
+    def ll_handle(self) -> Callable:
+        def fun(x, eta_loc, theta_scale, xh_scale):
+            eta_scale = np.matmul(xh_scale, theta_scale)
+            scale = self.model.inverse_link_scale(eta_scale)
+            loc = self.model.inverse_link_loc(eta_loc)
+            return ll(scale, loc, x)
+
+        return fun
```
Comment (picciama) [inline batchglm/train/numpy/glm_norm/model_container.py:81]: do we have some abstract properties inherited from the base `model_container` that should be added here which raise `NotImplementedError` like it is done for the poisson model?

File: batchglm/train/numpy/glm_norm/utils.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import math
+import numpy as np
+
+def ll(scale, loc, x):
```
Comment (picciama) [inline batchglm/train/numpy/glm_norm/utils.py:4]: shouldn't this be in `model_container`?

File: tests/numpy/test_accuracy_extreme_values.py [hunk 10]
```diff
@@ -35,7 +35,7 @@ class TestAccuracyXtremeNb(_TestAccuracyXtremeAll):
     """
 
     def test_nb(self) -> bool:
-        logger.error("TestAccuracyXtremeNb.test_nb()")
+        logger.error("TestAccuracyXtremeNb.test_nb() HERE!!")
```
Comment (picciama) [inline tests/numpy/test_accuracy_extreme_values.py:38]: remove `HERE`

File: tests/numpy/test_accuracy_extreme_values.py [hunk 11]
```diff
@@ -52,10 +52,10 @@ def test_norm(self) -> bool:
...
         logger.info("Normal noise model not implemented for numpy")
 
-        # np.random.seed(1)
-        # self._test_low_values(noise_model="norm")
-        # self._test_zero_variance(noise_model="norm")
-        return True
+        np.random.seed(1)
+        ret_val = self._test_low_values(noise_model="norm")
+        np.random.seed(1)
+        return ret_val and self._test_zero_variance(noise_model="nb")
```
Comment (picciama) [inline tests/numpy/test_accuracy_extreme_values.py:58]: the tests shouldn't actually have a return value but instead `assert False` if the fail.

**Review by picciama — CHANGES_REQUESTED** [review]
Seems as if there is a bit more to do. I get that this is different to the other models. Maybe we should have a chat about that with David again. Especially concerning the training procedure and close form.

File: batchglm/train/numpy/glm_norm/estimator.py (hunk 6 above)
Comment (ilan-gold) [reply batchglm/train/numpy/glm_norm/estimator.py:46]: Yes I added a warning about closed form for location.  If we fit the model there, then `train` doesn't actually have to do anything.

File: batchglm/train/numpy/glm_norm/utils.py (hunk 9 above)
Comment (ilan-gold) [reply batchglm/train/numpy/glm_norm/utils.py:4]: I guess but it's not a class method.  I could move it there though.

**Review by picciama — APPROVED** [review]

File: tests/numpy/test_accuracy_extreme_values.py (hunk 11 above)
Comment (picciama) [reply tests/numpy/test_accuracy_extreme_values.py:56]: yes do that please, I removed the returns for the nb related stuff already on my branches

File: tests/numpy/test_accuracy_extreme_values.py (hunk 10 above)
Comment (picciama) [reply tests/numpy/test_accuracy_extreme_values.py]: done?

File: batchglm/train/numpy/glm_norm/utils.py (hunk 9 above)
Comment (picciama) [reply batchglm/train/numpy/glm_norm/utils.py]: we should make it consistent though. Is this resolved?

File: batchglm/train/numpy/glm_norm/model_container.py (hunk 8 above)
Comment (picciama) [reply batchglm/train/numpy/glm_norm/model_container.py:125]: ok, then if you think we don't need that you can resolve this comment

File: batchglm/train/numpy/glm_norm/estimator.py (hunk 7 above)
Comment (picciama) [reply batchglm/train/numpy/glm_norm/estimator.py:75]: if you say it makes sense, leave it like that. Just make sure we have it properly documented and unit tests are covering this.

File: batchglm/train/numpy/glm_norm/estimator.py (hunk 6 above)
Comment (picciama) [reply batchglm/train/numpy/glm_norm/estimator.py:45]: ok, mark as resolved then I guess?

File: batchglm/train/numpy/glm_norm/estimator.py (hunk 5 above)
Comment (picciama) [reply batchglm/train/numpy/glm_norm/estimator.py]: is that done? It's important to import everything from the external.py so that we have consistent imports and standalone submodules that could easiyl be integrated elsewhere and moved

File: batchglm/models/glm_norm/utils.py (hunk 4 above)
Comment (picciama) [reply batchglm/models/glm_norm/utils.py:114]: ok, mark as resolved then I guess?

**Review by picciama — COMMENTED** [review]
Please resolve any comment and merge once done or request another approval.

File: batchglm/train/numpy/glm_norm/estimator.py (hunk 5 above)
Comment (ilan-gold) [reply batchglm/train/numpy/glm_norm/estimator.py]: yes!

File: batchglm/train/numpy/glm_norm/utils.py (hunk 9 above)
Comment (ilan-gold) [reply batchglm/train/numpy/glm_norm/utils.py]: yes it's in that now.

File: tests/numpy/test_accuracy_extreme_values.py (hunk 10 above)
Comment (ilan-gold) [reply tests/numpy/test_accuracy_extreme_values.py]: yes!

File: tests/numpy/test_accuracy_extreme_values.py (hunk 11 above)
Comment (ilan-gold) [reply tests/numpy/test_accuracy_extreme_values.py:56]: @picciama I have had a look and I simply think checking whether the model actually fits or not is the only goal.   The underlying class seems to be just from the `test_accuracy.py` file, which looks at accuracy.  But at the top of this file, it pretty clearly states "Test whether numerical extremes throw error in initialisation or during first training steps." and then the subsequent classes say something like "Test whether optimizers yield exact results for Poisson distributed data." which seems to be at odds with the goal of the top (or even be a goal that makes sense - these extreme value tests shouldn't be fitting anything, right?)

File: batchglm/models/glm_norm/utils.py (hunk 1 above)
Comment (ilan-gold) [reply batchglm/models/glm_norm/utils.py:82]: yup.

File: tests/numpy/test_accuracy_extreme_values.py (hunk 11 above)
Comment (picciama) [reply tests/numpy/test_accuracy_extreme_values.py:56]: True, that is misleading. I had a look at it. My suggestion would be to test for initialisation as well as accuracy after training. Accuracy used to be a problem as we had parameter clipping margins that prevented the model from being accurate enough in case for very small or large counts. I removed them for that reason / let them fail. Since I found a way around this while implementing edgeR, we can readd the accuracy testing at a later point after fixing the parameter clipping, since this needs thorough testing as well (should be done on separate branch). 
Therefore, you can leave it like that for now. We can also replace the return with assert later on, this isn't to big of an issue.
